client/sftp_client.py

```diff
@@ -196,202 +614,749 @@ def load_dirs(self, path, parent=None):
-		self.builder.add_from_file(target_file)
+		self.builder.add_from_file(gtk_builder_file)
 		self.window = self.builder.get_object('SFTPClientGUI.window')
-		self.application = application
-		upload_button = self.builder.get_object('button_upload')
-		download_button = self.builder.get_object('button_download')
-		self.logger = Logger(self.builder)
-		self.local = LocalDirectory(self.builder, self.logger, upload_button, self.application)  # pylint: disable=unused-variable
-		self.remote = RemoteDirectory(self.builder, self.logger, download_button, self.application, ftp)  # pylint: disable=unused-variable
-		self.local.upload_button.connect('button-press-event', self.upload)
-		self.remote.download_button.connect('button-press-event', self.download)
-		self.acceptable_perms = ('r-x', 'r--', 'rwx', 'rw-')
+		self.menubar = self.builder.get_object('SFTPClientGUI.menu')
+		self.render_menubar(self.menubar)
+		self.status_display = StatusDisplay(self.builder, self.queue)
+		self.local = LocalDirectory(self.builder, self.application)
+		self.remote = RemoteDirectory(self.builder, self.application, ftp, ssh)
+		self.builder.get_object('button_upload').connect('button-press-event', lambda widget, event: self._queue_transfer(UploadTask))
+		self.builder.get_object('button_download').connect('button-press-event', lambda widget, event: self._queue_transfer(DownloadTask))
+		self.local.menu_item_transfer.connect('activate', lambda widget: self._queue_transfer(UploadTask))
+		self.remote.menu_item_transfer.connect('activate', lambda widget: self._queue_transfer(DownloadTask))
+
+		box_local = self.builder.get_object('box_local_dropdown')
+		self._local_dropdown_model = Gtk.ListStore(str)
+		self.render_dropdown(self._local_dropdown_model, self.local)
+		self.local_dropdown = Gtk.ComboBox.new_with_model_and_entry(self._local_dropdown_model)
+		self.local_dropdown.set_entry_text_column(0)
+		self.local_dropdown.connect('changed', self.signal_combo_changed, self.local)
+		box_local.pack_start(self.local_dropdown, True, True, 0)
+
+		box_remote = self.builder.get_object('box_remote_dropdown')
+		self._remote_dropdown_model = Gtk.ListStore(str)
+		self.render_dropdown(self._remote_dropdown_model, self.remote)
+		self.remote_dropdown = Gtk.ComboBox.new_with_model_and_entry(self._remote_dropdown_model)
+		self.remote_dropdown.set_entry_text_column(0)
+		self.remote_dropdown.connect('changed', self.signal_combo_changed, self.remote)
+		box_remote.pack_start(self.remote_dropdown, True, True, 0)
+
+		self.window.connect('destroy', self.signal_window_destroy)
 		self.window.show_all()
 
-	def upload(self, temp, temp1):
-		selection = self.local.treeview_local.get_selection()
+
+	def signal_combo_changed(self, combo, *data):
+		system = data[0]
+		treeiter = combo.get_active_iter()
+		model = combo.get_model()
+		if treeiter != None:
+			new_dir = model[treeiter][0]
+		else:
+			# the user has typed something into the combo box
+			new_dir = '.'
```

Comment:
Make dot and dot dot into constants defined at the top of the plugin, then refer to them by name.